source/common/network/lc_trie.h

```diff
@@ -0,0 +1,588 @@
+#pragma once
+
+#include <arpa/inet.h>
+
+#include <algorithm>
+#include <climits>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/network/address.h"
+
+#include "common/common/assert.h"
+#include "common/common/empty_string.h"
+#include "common/network/cidr_range.h"
+#include "common/network/utility.h"
+
+#include "absl/numeric/int128.h"
+#include "fmt/format.h"
+
+namespace Envoy {
+namespace Network {
+namespace LcTrie {
+
+/**
+ * Level Compressed Trie for tagging IP addresses. Both IPv4 and IPv6 addresses are supported
+ * within this class with no calling pattern changes.
+ *
+ * The algorithm to build the LC-Trie is desribed in the paper 'IP-address lookup using LC-tries'
+ * by'S. Nilsson' and 'G. Karlsson'. Refer to LcTrieInternal for implementation and algorithm
```

Comment:
nit: missing space after "by". Please also provide direct link to paper/article if possible.

Reply:
Fixed. Found a free version of the paper hosted by the author.